src/main/java/org/elasticsearch/common/rounding/TimeZoneRounding.java

```diff
@@ -147,21 +122,36 @@ public byte id() {
 
         @Override
         public long roundKey(long utcMillis) {
-            long offset = preTz.getOffset(utcMillis);
-            long time = utcMillis + offset;
-            return field.roundFloor(time) - offset;
+            long timeLocal = utcMillis;
+            if (!isUtc) {
+                timeLocal = timeZone.convertUTCToLocal(utcMillis);
+            }
+            long rounded = field.roundFloor(timeLocal);
+            if (!isUtc) {
+                return timeZone.convertLocalToUTC(rounded, true, utcMillis);
```

Comment:
@cbuescher Making performance tests to see whether we would benefit from moving some computations to valueForKey would be awesome, but I'm already happy that we are making the time zone handling more reasonable, so I would be totally fine with investigating performance in a different issue/PR.